types/underscore/index.d.ts

```diff
@@ -890,15 +929,14 @@ declare module _ {
         compact<T>(array: _.List<T | null | undefined | false | "" | 0> | null | undefined): T[]
 
         /**
-        * Flattens a nested array (the nesting can be to any depth). If you pass shallow, the array will
-        * only be flattened a single level.
-        * @param array The array to flatten.
-        * @param shallow If true then only flatten one level, optional, default = false.
-        * @return `array` flattened.
-        **/
-        flatten(
-            array: _.List<any>,
-            shallow?: boolean): any[];
+         * Flattens a nested array (the nesting can be to any depth). If you pass shallow, the array will
+         * only be flattened a single level.
+         * @param list The array to flatten.
+         * @param shallow If true then only flatten one level, optional, default = false.
+         * @returns The flattened list.
+         **/
+        flatten<T>(list: List<T>, shallow?: false): DeepFlattenedList<T>[];
+        flatten<T>(list: List<T>, shallow: true): ShallowFlattenedList<T>[];
```

Comment:
That's true, good point, if I can drop `& object` I'll definitely replace this with `ListItem`.

Reply:
Yup, the safety guard of `& object` isn't needed for shallow flattening because unlike deep flattening it doesn't recurse its way to `any[]`. On the other hand, I spoke too soon, `ListItem` and `ShallowFlattenedList` differ slightly in the else parts of their ternary expressions since `ListItem` yields `never` and `ShallowFlattenedList` yields the original type since one should still technically be able to call `flatten` on a single-dimension array and get back that same type.

If you'd prefer to block the scenario of flattening an already flat list since it's not useful, I could switch to `ListItem` if you'd like, but that technically won't be accurate form the perspective of actual JS behavior.

Reply:
No you are right that it is not the same as the pre-existing `ListItem` type. I was being sloppy; I didn't mean to refer to that other type, I just thought that `ListItem` was a more appropriate name when considering this type in isolation.

I don't like the name `ShallowFlattenedListItem` because `T` doesn't need to have any flattening history in order for this type to make sense. However, it's hard to come up with a better name. Maybe `OptionalListItem`? And then `DeepestListItem` for `DeepFlattenedList`?

Reply:
I like `DeepestListItem`. I'd be ok with `OptionalListItem`, but how do you feel about `ListItemOrType` as a potential alternative? My only worry is that `Optional` personally makes me think along the lines of optional chaining where I'd be possibly be getting back `undefined`.

Reply:
Yeah I agree `Optional` is not a great choice here. `Type` is never terribly informative in a type name, but maybe `ListItemOrSelf`?

Reply:
Ooh, I like it, thanks!

Reply:
Also I owe you an apology, it turns out I was the one being sloppy in the `ListItem` discussion because I was thinking `ListItem` was what `TypeOfList` was called. `ListItem` is an available type name, but I think I still prefer `ListItemOrSelf` since `ListItem` still makes me think I'll get something like `never` or `undefined` back if the type is not a list.